### ISSUE TITLE
observation_and_action_constraint_splitter with LinearUCBAgent using MovieLensPyEnvironment

### ISSUE THREAD

**Issue by My3VM** [tensorflow/agents#822]
Hello all,

I have been trying to mask some of the actions using the observation_and_action_constraint_splitter property of the LinearUCBAgent with intentions to disable certain arms after they are being pulled.
NUM_ACTIONS is currently set to 20

While this configuration seemingly look ok with Python function (@tf.function which is not able to decode expected TensorSpec datatype returned from the function)

**#flavor a**
def observation_and_action_constraint_splitter(observation):
    print('observation ',observation)
    action_mask = np.ones(NUM_ACTIONS).astype(dtype=np.int32)
    tensor_action_mask = tf.convert_to_tensor(action_mask, dtype=tf.float32)
    tensor_action_mask = tf.TensorSpec.from_tensor(tensor_action_mask)
    return observation, tensor_action_mask

agent = lin_ucb_agent.LinearUCBAgent(
        time_step_spec=environment.time_step_spec(),
        action_spec=environment.action_spec(),
        tikhonov_weight=0.001,
        alpha=AGENT_ALPHA,
        dtype=tf.float32,
        accepts_per_arm_features=FLAGS.per_arm,
        observation_and_action_constraint_splitter=observation_and_action_constraint_splitter)

With this config, the training runs 2 steps and immediately starts receiving a scalar value as observation, causing issues with mismatching datatypes downstream.
BATCH_SIZE  1
I0213 14:57:06.110655 139726866007872 trainer.py:130] Did not find a pre-existing checkpoint. Starting from scratch.
observation  **TensorSpec**(shape=(20,), dtype=tf.float64, name='observation')
observation  **TensorSpec**(shape=(20,), dtype=tf.float64, name='observation')
observation  **Tensor**("observation:0", shape=(None, 20), dtype=float64)
observation  **Tensor**("observation:0", shape=(None, 20), dtype=float64)
failed:  Shapes (None, 20) and (20,) are incompatible

I have also tried giving static data as tensorSpec from observation_and_action_constraint_splitter, while ensuring it always returns the dims are retained, still it ends up with dtype erros after 2 steps:
**#flavor b**
def observation_and_action_constraint_splitter(observation):
    #print('observation ',observation)
    vals = [[-3.03152584e-01 , 5.71876601e-01, -5.63917888e-02,  1.62131316e-01,
   1.08664307e-01,  3.49876896e-01,  2.57143517e-01,  5.41114355e-02,
  -3.87893791e-01,  2.59026888e-01,  1.25500056e-01,  8.51947348e-02,
   1.51225168e-01,  6.41737045e-02,  1.02322254e-01,  3.25405081e-04,
  -4.77013915e-02, -2.42873893e-02,  3.22570274e-02,  5.73024425e-03]]
    obs = tf.convert_to_tensor(vals, dtype=tf.float64)
    print('obs ',obs)
    tensor_obs = tf.TensorSpec.from_tensor(obs)
    print('tensor_obs ',tensor_obs)
    action_mask = np.ones(NUM_ACTIONS).astype(dtype=np.int32)
    tensor_action_mask = tf.convert_to_tensor(action_mask, dtype=tf.float32)
    tensor_action_mask = tf.TensorSpec.from_tensor(tensor_action_mask)
    return tensor_obs, tensor_action_mask

Error logs after 2 steps on training:
obs  tf.Tensor(
[[-3.03152584e-01  5.71876601e-01 -5.63917888e-02  1.62131316e-01
   1.08664307e-01  3.49876896e-01  2.57143517e-01  5.41114355e-02
  -3.87893791e-01  2.59026888e-01  1.25500056e-01  8.51947348e-02
   1.51225168e-01  6.41737045e-02  1.02322254e-01  3.25405081e-04
  -4.77013915e-02 -2.42873893e-02  3.22570274e-02  5.73024425e-03]], shape=(1, 20), dtype=float64)
tensor_obs  TensorSpec(shape=(1, 20), dtype=tf.float64, name=None)
obs  tf.Tensor(
[[-3.03152584e-01  5.71876601e-01 -5.63917888e-02  1.62131316e-01
   1.08664307e-01  3.49876896e-01  2.57143517e-01  5.41114355e-02
  -3.87893791e-01  2.59026888e-01  1.25500056e-01  8.51947348e-02
   1.51225168e-01  6.41737045e-02  1.02322254e-01  3.25405081e-04
  -4.77013915e-02 -2.42873893e-02  3.22570274e-02  5.73024425e-03]], shape=(1, 20), dtype=float64)
tensor_obs  TensorSpec(shape=(1, 20), dtype=tf.float64, name=None)
obs  tf.Tensor(
[[-3.03152584e-01  5.71876601e-01 -5.63917888e-02  1.62131316e-01
   1.08664307e-01  3.49876896e-01  2.57143517e-01  5.41114355e-02
  -3.87893791e-01  2.59026888e-01  1.25500056e-01  8.51947348e-02
   1.51225168e-01  6.41737045e-02  1.02322254e-01  3.25405081e-04
  -4.77013915e-02 -2.42873893e-02  3.22570274e-02  5.73024425e-03]], shape=(1, 20), dtype=float64)
tensor_obs  TensorSpec(shape=(1, 20), dtype=tf.float64, name=None)
**failed:  Failed to convert elements of TensorSpec(shape=(1, 20), dtype=tf.float64, name=None) to Tensor. Consider casting elements to a supported type. See https://www.tensorflow.org/api_docs/python/tf/dtypes for supported TF dtypes**

Any hints /tips towards understanding this behaviour would be of great help.

**Comment by My3VM** [comment]
Could resolve this issue by embedding action mask tensor in 2d array and returning tensor directly, instead of the tensorspec.
